src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -2038,13 +2058,80 @@ MulticopterPositionControl::task_main()
 				}
 			}
 
-			/* control roll and pitch directly if no aiding velocity controller is active */
 			if (!_control_mode.flag_control_velocity_enabled) {
 				_att_sp.roll_body = _manual.y * _params.man_roll_max;
 				_att_sp.pitch_body = -_manual.x * _params.man_pitch_max;
 
-				/* only if optimal recovery is not used, modify roll/pitch */
-				if (_params.opt_recover <= 0) {
+				/* If ACRO attitude control mode */
+				if (_vehicle_status.nav_state == vehicle_status_s::NAVIGATION_STATE_ACRO) {
+
+					float rollRate = _manual.y * _params.acro_rollRate_max;
+					float pitchRate = -_manual.x * _params.acro_pitchRate_max;
+					float yawRate = _manual.r * _params.acro_yawRate_max;
+
+					// attitude setpoint quaternion
+					matrix::Quatf R_sp(_att_sp.q_d);
+
+
+					/* the control sequencer overrides manual inputs
+					 * by setting values for roll/pitch/yawRate and modifying R_sp as needed
+					 */
+					if (!_sequencer_initialized) {
+						_sequencer_initialized = true;
+						prog_sequence_init(sequence_set::coord_turn, 25.0f);
+					}
+
+					prog_sequence(_ctrl_state, _att_sp, _manual, R_sp, rollRate, pitchRate, yawRate);
+
+					float tilt_error = 0.0f;
+					float yaw_error = 0.0f;
+
+#undef LIMIT_LEAD_ANGLE
+#ifdef LIMIT_LEAD_ANGLE
+					/* limit setpoint lead angle */
+					math::Quaternion q_cur(_ctrl_state.q);
+					math::Matrix<3, 3> R_cur = q_cur.to_dcm();
+					math::Vector<3> zb(R_cur.data[2]);
+					math::Vector<3> zsp(R_sp.data[2]);
+					tilt_error = acosf(zb * zsp);
+					math::Vector<3> xb(R_cur.data[0]);
+					math::Vector<3> xsp(R_sp.data[0]);
+					yaw_error = acosf(xb * xsp);
```

Comment:
This needs more commenting, I think. What is zb, zsp, xb, xsp and how do you get to the calculations. A reader without too much knowledge has no idea.